AAXtoMP3

```diff
@@ -1,166 +1,305 @@
+                      # Extremely verbose output.
+    -d | --debug      ) DEBUG=1;                                                    shift ;;        
+                      # Command synopsis.
+    -h | --help       ) printf "$usage" $0 ;                                        exit ;;         
+                      # Standard flag signifying the end of command line processing.
+    --                )                                                             shift; break ;; 
+                      # Anything else stops command line processing.
+    *                 )                                                             break ;;        
+
+  esac
+done
+
+# -----
+# Empty argv means we have nothing to do so lets bark some help.
 if [ "$#" -eq 0 ]; then
-        echo "Usage: bash AAXtoMP3 [--flac] [--aac] [--opus ] [--single] AUTHCODE {FILES}"
-        echo "	Note that order does matter!"
-        exit 1
+  printf "$usage" $0
+  exit 1
 fi
 
-if [[ "$1" = '--flac' ]]
-then
-    codec=flac
-    extension=flac
-    shift
+# ========================================================================
+# Variable validation
+set -o errexit -o noclobber -o nounset -o pipefail
+
+# -----
+# Detect which annoying version fo grep we have
+GREP=$(grep --version | grep -q GNU && echo "grep" || echo "ggrep")
+if ! [[ $(type -P "$GREP") ]]; then
+  echo "$GREP (GNU grep) is not in your PATH"
+  echo "Without it, this script will break."
+  echo "On macOS, you may want to try: brew install grep"
+  exit 1
 fi
 
-if [[ "$1" == '--aac' ]]
-then
-    codec=copy
-    extension=m4a
-    mode=single
-    shift
+# -----
+# Detect if we need mp4art for cover additions to m4a & m4b files.
+if [[ "x${extension}" == "xm4a" && "x$(type -P mp4art)" == "x" ]]; then
+  echo "WARN mp4art was not found on your env PATH variable"
+  echo "Without it, this script will not be able to add cover art to"
+  echo "m4b files. Note if there are no other errors the AAXtoMP3 will"
+  echo "continue. However no cover art will be added to the output."
+  echo "INSTALL:" 
+  echo "MacOS:   brew install mp4v2"
+  echo "Ubuntu:  sudo apt-get install mp4v2-utils"
 fi
 
-if [[ "$1" = '--opus' ]]
-then
-    codec=libopus
-    extension=ogg
-    shift
+# -----
+# Obtain the authcode from either the command line,  local directory or home directory.
+# See Readme.md for details on how to aquire your personal authcode for your personal
+# audible AAX files.
+if [ -z $auth_code ]; then
+  if [ -r .authcode ]; then
+    auth_code=`head -1 .authcode`
+  elif [ -r ~/.authcode ]; then
+    auth_code=`head -1 ~/.authcode`
+  fi
+fi
+# No point going on if no authcode found.
+if [ -z $auth_code ]; then
+  echo "ERROR Missing authcode"
+  echo "$usage"
+  exit 1  
 fi
 
-if [[ "$1" == '--single' ]]
-then
-    mode=single
-    shift
+# -----
+# Check the target dir for if set if it is writable
+if [[ ! -w "${targetdir}" || ! -d "${targetdir}" ]] ; then
```

Comment:
This check should after https://github.com/jostyee/AAXtoM4B/pull/3/files#diff-f1655c3b19357414d6913f2a6bdc7f37R197 , which set a default output directory, otherwise when don't pass `-t` the script would fail.

Reply:
Yep there is a bug there.   I added a check to see if the -t option was passed.  If it was then it checks for the target directory existence and is writable.